lightning/src/ln/interactivetxs.rs

```diff
@@ -290,33 +315,86 @@ impl NegotiationContext {
 		}
 	}
 
-	fn sent_tx_add_input(&mut self, msg: &msgs::TxAddInput) {
+	fn sent_tx_add_input(&mut self, msg: &msgs::TxAddInput) -> Result<(), AbortReason> {
 		let tx = msg.prevtx.as_transaction();
 		let input = TxIn {
 			previous_output: OutPoint { txid: tx.txid(), vout: msg.prevtx_out },
 			sequence: Sequence(msg.sequence),
 			..Default::default()
 		};
-		debug_assert!((msg.prevtx_out as usize) < tx.output.len());
-		let prev_output = &tx.output[msg.prevtx_out as usize];
-		self.prevtx_outpoints.insert(input.previous_output.clone());
-		self.inputs.insert(
-			msg.serial_id,
-			TxInputWithPrevOutput { input, prev_output: prev_output.clone() },
-		);
+		let prev_output =
+			tx.output.get(msg.prevtx_out as usize).ok_or(AbortReason::PrevTxOutInvalid)?.clone();
+		if !self.prevtx_outpoints.insert(input.previous_output.clone()) {
+			// We have added an input that already exists
+			return Err(AbortReason::PrevTxOutInvalid);
+		}
+		self.inputs.insert(msg.serial_id, TxInputWithPrevOutput { input, prev_output });
+		Ok(())
 	}
 
-	fn sent_tx_add_output(&mut self, msg: &msgs::TxAddOutput) {
+	fn sent_tx_add_output(&mut self, msg: &msgs::TxAddOutput) -> Result<(), AbortReason> {
 		self.outputs
 			.insert(msg.serial_id, TxOut { value: msg.sats, script_pubkey: msg.script.clone() });
+		Ok(())
 	}
 
-	fn sent_tx_remove_input(&mut self, msg: &msgs::TxRemoveInput) {
+	fn sent_tx_remove_input(&mut self, msg: &msgs::TxRemoveInput) -> Result<(), AbortReason> {
 		self.inputs.remove(&msg.serial_id);
+		Ok(())
 	}
 
-	fn sent_tx_remove_output(&mut self, msg: &msgs::TxRemoveOutput) {
+	fn sent_tx_remove_output(&mut self, msg: &msgs::TxRemoveOutput) -> Result<(), AbortReason> {
 		self.outputs.remove(&msg.serial_id);
+		Ok(())
+	}
+
+	fn check_counterparty_fees(
+		&self, counterparty_inputs_value: u64, counterparty_outputs_value: u64,
+	) -> Result<(), AbortReason> {
+		let mut counterparty_weight_contributed: u64 = self
+			.counterparty_outputs_contributed()
+			.map(|output| get_output_weight(&output.script_pubkey))
+			.sum();
+		// We don't know the counterparty's witnesses ahead of time obviously, so we need to have some
+		// estimates for their weights based on script type. We make some reasonably assumptions here.
```

Comment:
Yeah I think this commit needs more work.

The order for sending signatures is strictly the peer with lowest total input value first (unless equal then I believe it is initiator). This logic is in #2302 already but I'm in the process of rebasing on this PR.

Reply:
Well I actually wanted to bring up the idea of a peer providing the witness weight in a TLV of the tx_add_input message. I'm not sure how practical that is and if it's easy for implementations to have that information up front (I know for us it could be tricky at least).

I've been looking at how other implementations do it and it it seems it usually involves margins of error and sometimes just being happy overpaying in some situations.

Reply:
Would that be a private protocol extension by LDK? It makes sense, though it can be only treated as a hint, as it could  be missing or it coculd be plain wrong.

Reply:
I was considering a different idea: once we have the full signed transactions, we could check the actual fee against the planned fee rate, and at least log if lower, or maybe error if significantly lower. At this point the negotiation is over, so cannot be simply aborted via `tx_abort`.
In the rare case of low actual fees, due to e.g. much larger actual witness, what happens is that we broadcast the transaction and wait, and the tx not getting confirmed for very long time, or never if it fails to reach even the purging rate. It's very hard to know what happened in such a case.

Reply:
>  and the tx not getting confirmed for very long time, or never if it fails to reach even the purging rate. It's very hard to know what happened in such a case.

We'd need to inform the user to double spend one of the inputs.

I've left some updates. The spec says just consider the lower bound for weights as in BOLT 3 for P2WPKH, P2WSH, etc. Then we will do the pre-check and afterward once both parties have signed we can check the actual feerate against target as you said with a 5% margin of error or so. If it falls below this margin then we can inform the user it's probably a good idea to try double spend. But either way we will still need to persist the channel as the counterparty might broadcast anyway.

Reply:
Yeah I think I agree with that again. 
Contributing liquidity also involves some level of trust in the peer. So I feel like we're doing extra work and we're still going to have some holes around pinning as you said. 

So we'll just keep the checks basic on the unsigned tx here and then on the fully signed tx, we can do the feerate check with a little wiggle room.